### PR TITLE
feat: Improve readme

### PR DESCRIPTION
**Pull request by ksimon1**
**What this PR does / why we need it**:
Main purpose of this change is to improve readme of tasks and insert task runs
inside it. This will help users when they access tasks via artifact hub.
This commit also generates new taskruns with included resolver -
https://tekton.dev/docs/pipelines/hub-resolver/#task-resolution. That
means users doesn't have to download tasks into clusters, because tekton
will download it automatically with informations from resolver

**Release note**:
```
Insert taskruns inside README

```

### CODE REVIEW

File: templates/copy-template/readmes/README.md [hunk 1]
```diff
@@ -19,7 +18,16 @@ A bundle of predefined templates to use can be found in [Common Templates](https
 
 ### Usage
 
-Please see [examples](examples) on how to copy a template.
+Please see examples on how to copy a template:
+Task run:
+```
+{{ task_run_yaml | to_nice_yaml }}
```
Comment (jcanocan) [inline templates/copy-template/readmes/README.md:24]: What is expected to users to apply here? A yaml manifest like the previous ones? If that the case, please add it here or reference it?

File: templates/create-vm-from-manifest/readmes/README.md [hunk 2]
```diff
@@ -20,7 +20,16 @@ This task creates a VirtualMachine from YAML manifest
 
 ### Usage
 
-Please see [examples](examples) on how to create VMs.
+Please see examples on how to create VMs:
```
Comment (jcanocan) [inline templates/create-vm-from-manifest/readmes/README.md:23]:
```suggestion
Task run:
```

File: templates/create-vm-from-template/readmes/README.md [hunk 3]
```diff
@@ -24,7 +24,16 @@ A bundle of predefined templates to use can be found in [Common Templates](https
 
 ### Usage
 
-Please see [examples](examples) on how to create VMs from a template.
+Please see examples on how to create VMs from a template:
```
Comment (jcanocan) [inline templates/create-vm-from-template/readmes/README.md:27]:
```suggestion
Task run:
```

File: templates/disk-virt-customize/readmes/README.md [hunk 4]
```diff
@@ -10,7 +10,16 @@ This task uses [virt-customize](https://libguestfs.org/virt-customize.1.html) to
 
 ### Usage
 
-Please see [examples](examples)
+Please see examples:
```
Comment (jcanocan) [inline templates/disk-virt-customize/readmes/README.md:13]:
```suggestion
Task run:
```

File: templates/copy-template/readmes/README.md (hunk 1 above)
Comment (jcanocan) [reply templates/copy-template/readmes/README.md:24]: Please, check also the rest of these examples.

File: templates/disk-virt-sysprep/readmes/README.md [hunk 5]
```diff
@@ -11,7 +11,16 @@ This task uses [virt-sysprep](https://libguestfs.org/virt-sysprep.1.html) to run
 
 ### Usage
 
-Please see [examples](examples)
+Please see examples:
```
Comment (jcanocan) [inline templates/disk-virt-sysprep/readmes/README.md:14]:
```suggestion
Task run:
```

File: templates/execute-in-vm/readmes/README.md [hunk 6]
```diff
@@ -37,19 +37,37 @@ Following secret data keys are recognized for SSH connections:
 - **disable-strict-host-key-checking**: host-public-key (authorized-key) does not have to be supplied when this value is set to true.
 - **additional-ssh-options**: Additional arguments to pass to the SSH command.
 
-Please see [secret](examples/secrets) examples.
+Please see secrets examples:
```
Comment (jcanocan) [inline templates/execute-in-vm/readmes/README.md:40]:
```suggestion
Task run:
```

File: templates/execute-in-vm/readmes/README.md [hunk 7]
```diff
@@ -37,19 +37,37 @@ Following secret data keys are recognized for SSH connections:
...
+```
+{{ ssh_secret_yaml | to_nice_yaml }}
+```
 
 ### Usage
 
-Please see [examples](examples).
-
-#### Specific examples
+Please see examples:
```
Comment (jcanocan) [inline templates/execute-in-vm/readmes/README.md:47]:
```suggestion
```

File: templates/generate-ssh-keys/readmes/README.md [hunk 8]
```diff
@@ -20,7 +20,16 @@ This task uses `ssh-keygen` to generate a private and public key pair
 
 ### Usage
 
-Please see [examples](examples)
+Please see examples:
```
Comment (jcanocan) [inline templates/generate-ssh-keys/readmes/README.md:23]:
```suggestion
Task run:
```

File: templates/modify-data-object/readmes/README.md [hunk 9]
```diff
@@ -16,7 +16,16 @@ This task modifies a data object (DataVolumes or DataSources).
 
 ### Usage
 
-Please see [examples](examples) on how to modify data objects.
+Please see examples on how to modify data objects:
```
Comment (jcanocan) [inline templates/modify-data-object/readmes/README.md:19]:
```suggestion
Task run:
```

File: templates/modify-vm-template/readmes/README.md [hunk 10]
```diff
@@ -24,7 +24,16 @@ Please see [RBAC permissions for running the tasks](../../docs/tasks-rbac-permis
 
 ### Usage
 
-Please see [examples](examples) on how to do a copy template from a template.
+Please see examples on how to do a copy template from a template:
```
Comment (jcanocan) [inline templates/modify-vm-template/readmes/README.md:27]:
```suggestion
Task run:
```

File: templates/modify-windows-iso-file/readmes/README.md [hunk 11]
```diff
@@ -13,5 +13,17 @@ bootloader will not require any key pres.
 
 ### Usage
 
-Please see [examples](examples) on how to run iso modification task.
-The task run has to specify spec.podTemplate.securityContext! See [examples](examples) for example how to specify it.
+Please see examples on how to run iso modification task:
```
Comment (jcanocan) [inline templates/modify-windows-iso-file/readmes/README.md:16]:
```suggestion
Task run:
```

File: templates/wait-for-vmi-status/readmes/README.md [hunk 12]
```diff
@@ -14,7 +14,16 @@ This task waits for a specific status of a VirtualMachineInstance (VMI) and fail
 
 ### Usage
 
-Please see [examples](examples)
+Please see examples:
```
Comment (jcanocan) [inline templates/wait-for-vmi-status/readmes/README.md:17]:
```suggestion
Task run:
```

**Review by jcanocan — COMMENTED** [review]
Great effort! Thanks.
As general comment, I see that in some places we are using `tekton.dev/v1beta1` in others `tekton.dev/v1`. Should we be consistent and use either `v1` or `v1beta1` on all of them? Or it will break the expected behavior?

**Comment by ksimon1** [comment]
> Great effort! Thanks. As general comment, I see that in some places we are using `tekton.dev/v1beta1` in others `tekton.dev/v1`. Should we be consistent and use either `v1` or `v1beta1` on all of them? Or it will break the expected behavior?

This will be fixed in https://github.com/kubevirt/kubevirt-tekton-tasks/pull/331

**Comment by jcanocan** [comment]
> > Great effort! Thanks. As general comment, I see that in some places we are using `tekton.dev/v1beta1` in others `tekton.dev/v1`. Should we be consistent and use either `v1` or `v1beta1` on all of them? Or it will break the expected behavior?
> 
> This will be fixed in #331

That's great! Should we also update the versions of the examples? For instance:

https://github.com/kubevirt/kubevirt-tekton-tasks/pull/330/files#diff-c92e223f42529b4af35d7372f7146400d3d254c6d49b711d0fbb72d3357b851eR62

File: tasks/cleanup-vm/README.md [hunk 13]
```diff
@@ -37,16 +37,113 @@ Following secret data keys are recognized for SSH connections:
...
+```
+apiVersion: v1
+kind: Secret
+metadata:
+    name: ssh-secret
+stringData:
+    disable-strict-host-key-checking: 'true'
+    ssh-privatekey: '<private key should be here>
+
+        '
```
Comment (0xFelix) [inline tasks/cleanup-vm/README.md:50]: What are these lines for?

File: tasks/cleanup-vm/README.md [hunk 14]
```diff
@@ -37,16 +37,113 @@ Following secret data keys are recognized for SSH connections:
...
+        -   name: type
+            value: artifact
+        -   name: kind
+            value: task
+        -   name: name
+            value: cleanup-vm
+        -   name: version
+            value: v0.18.0
+        resolver: hub
+
```
Comment (0xFelix) [inline tasks/cleanup-vm/README.md:145]: nit: empty line

File: tasks/cleanup-vm/examples/secrets/ssh-secret.yaml [hunk 15]
```diff
@@ -8,5 +8,4 @@ stringData:
   disable-strict-host-key-checking: "true"
   ssh-privatekey: |
```
Comment (0xFelix) [inline tasks/cleanup-vm/examples/secrets/ssh-secret.yaml:9]:
```suggestion
  ssh-privatekey: |-
```

Could possible help to get rid of whitespace at the end.

File: tasks/copy-template/README.md [hunk 16]
```diff
@@ -2,8 +2,7 @@
 
 #### Task is deprecated and will be removed in future versions.
 
-This task copies a template.
-A bundle of predefined templates to use can be found in [Common Templates](https://github.com/kubevirt/common-templates) project.
+This task copies an KubeVirt Virtual Machine template. This task works only with Common Templates and on OKD / OpenShift. A bundle of predefined templates to use can be found in [Common Templates](https://github.com/kubevirt/common-templates) project.
```
Comment (0xFelix) [inline tasks/copy-template/README.md:5]: Is it really limited to common-templates. Wouldn't it work with any `Template`?

File: tasks/copy-template/README.md [hunk 17]
```diff
@@ -20,7 +19,52 @@ A bundle of predefined templates to use can be found in [Common Templates](https
...
+spec:
+    params:
+    -   name: sourceTemplateName
+        value: source-vm-template-example
+    -   name: targetTemplateName
+        value: target-vm-template-example
+    taskRef:
+        kind: Task
+        name: copy-template
+
```
Comment (0xFelix) [inline tasks/copy-template/README.md:38]: nit: newline

File: tasks/copy-template/README.md [hunk 18]
```diff
@@ -20,7 +19,52 @@ A bundle of predefined templates to use can be found in [Common Templates](https
...
+        -   name: type
+            value: artifact
+        -   name: kind
+            value: task
+        -   name: name
+            value: copy-template
+        -   name: version
+            value: v0.18.0
+        resolver: hub
+
```
Comment (0xFelix) [inline tasks/copy-template/README.md:66]: Same

File: tasks/copy-template/README.md [hunk 19]
```diff
@@ -61,3 +105,7 @@ subjects:
     name: copy-template-task
 ---
 ```
+
+### Platforms
+
+The Task can be run on linux/amd64 platform.
```
Comment (0xFelix) [inline tasks/copy-template/README.md:111]: Can we make it run on more platforms?

File: tasks/create-vm-from-manifest/README.md [hunk 20]
```diff
@@ -17,7 +17,70 @@ This task creates a VirtualMachine from YAML manifest
...
+            false\n  template:\n    metadata:\n      labels:\n        kubevirt.io/vm:
+            vm-cirros\n    spec:\n      domain:\n        devices:\n          disks:\n
+            \         - disk:\n              bus: virtio\n            name: containerdisk\n
+            \         - disk:\n              bus: virtio\n            name: cloudinitdisk\n
+            \       machine:\n          type: \"\"\n        resources:\n          requests:\n
+            \           memory: 64M\n      terminationGracePeriodSeconds: 0\n      volumes:\n
+            \     - containerDisk:\n          image: kubevirt/cirros-container-disk-demo\n
+            \       name: containerdisk\n      - cloudInitNoCloud:\n          userData:
+            |\n            #!/bin/sh\n            echo 'printed from cloud-init userdata'\n
+            \       name: cloudinitdisk"
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tasks/create-vm-from-manifest/README.md:41]: Use a multiline string perhaps?

File: tasks/create-vm-from-manifest/README.md [hunk 21]
```diff
@@ -17,7 +17,70 @@ This task creates a VirtualMachine from YAML manifest
...
+        -   name: type
+            value: artifact
+        -   name: kind
+            value: task
+        -   name: name
+            value: create-vm-from-manifest
+        -   name: version
+            value: v0.18.0
+        resolver: hub
+
```
Comment (0xFelix) [inline tasks/create-vm-from-manifest/README.md:82]: nit: empty line

File: tasks/create-vm-from-template/README.md [hunk 22]
```diff
@@ -22,7 +22,60 @@ A bundle of predefined templates to use can be found in [Common Templates](https
...
+        -   name: type
+            value: artifact
+        -   name: kind
+            value: task
+        -   name: name
+            value: create-vm-from-template
+        -   name: version
+            value: v0.18.0
+        resolver: hub
+
```
Comment (0xFelix) [inline tasks/create-vm-from-template/README.md:77]: Same

File: tasks/create-vm-from-template/README.md [hunk 23]
```diff
@@ -22,7 +22,60 @@ A bundle of predefined templates to use can be found in [Common Templates](https
...
+    -   name: templateParams
+        value:
+        - NAME:example-vm
+    -   name: ownDataVolumes
+        value:
+        - example-dv
+    taskRef:
+        kind: Task
+        name: create-vm-from-template
+
```
Comment (0xFelix) [inline tasks/create-vm-from-template/README.md:45]: Same

File: tasks/modify-data-object/README.md [hunk 24]
```diff
@@ -19,7 +19,58 @@ This task modifies a data object (DataVolumes or DataSources).
...
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+    name: modify-data-object-taskrun
+spec:
+    params:
+    -   name: waitForSuccess
+        value: 'true'
+    -   name: manifest
+        value: "apiVersion: cdi.kubevirt.io/v1beta1\nkind: DataVolume\nmetadata:\n
```
Comment (0xFelix) [inline tasks/modify-data-object/README.md:34]: Use multiline strings for manifests?

**Review by 0xFelix — COMMENTED** [review]
What about listing the examples with the task resolver as the first example? Or make the simple example used resolvers instead? Makes it even easier to use the tasks.

Also there a lot of empty lines at the end of code blocks. Can you try to remove them?

**Comment by ksimon1** [comment]
> What about listing the examples with the task resolver as the first example? Or make the simple example used resolvers instead? Makes it even easier to use the tasks.
> 

I was thinking about using the taskruns with resolvers as the only example. Your comment confirms, that I should do it :)

File: templates/copy-template/readmes/README.md (hunk 1 above)
Comment (ksimon1) [reply templates/copy-template/readmes/README.md:24]: I don't understand your comment. This is a jinja template which will be populated with variables by ansible during generating.

File: tasks/copy-template/README.md (hunk 16 above)
Comment (ksimon1) [reply tasks/copy-template/README.md:5]: Correct, copy template task can work with any template

File: tasks/copy-template/README.md (hunk 19 above)
Comment (ksimon1) [reply tasks/copy-template/README.md:92]: Not now.

File: tasks/cleanup-vm/README.md [hunk 25]
```diff
@@ -41,12 +41,45 @@ Please see [secret](examples/secrets) examples.
...
-#### Specific examples
-
-- [delete a VM](examples/taskruns/cleanup-vm-simple-taskrun.yaml)
-- [stop postgresql service over ssh and stop a VM](examples/taskruns/cleanup-vm-with-ssh-taskrun.yaml)
+Task run using resolver:
+```
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+    name: cleanup-vm-taskrun-resolver
```
Comment (0xFelix) [inline tasks/cleanup-vm/README.md:49]:
```suggestion
    generateName: cleanup-vm-taskrun-resolver-
```

To allow creating the TaskRun multiple times?

File: tasks/create-vm-from-manifest/README.md [hunk 26]
```diff
@@ -17,7 +17,73 @@ This task creates a VirtualMachine from YAML manifest
...
+                devices:
+                    disks:
+                    -   disk:
+                            bus: virtio
+                        name: containerdisk
+                    -   disk:
+                            bus: virtio
+                        name: cloudinitdisk
+                machine:
+                    type: ''
```
Comment (0xFelix) [inline tasks/create-vm-from-manifest/README.md:70]: Omit this, should be defaulted appropriately by mutators?

File: tasks/create-vm-from-manifest/README.md [hunk 27]
```diff
@@ -17,7 +17,73 @@ This task creates a VirtualMachine from YAML manifest
...
+                resources:
+                    requests:
+                        memory: 64M
```
Comment (0xFelix) [inline tasks/create-vm-from-manifest/README.md:73]: Use `domain.memory.guest` instead?

File: tasks/create-vm-from-manifest/README.md [hunk 28]
```diff
@@ -17,7 +17,73 @@ This task creates a VirtualMachine from YAML manifest
...
+            terminationGracePeriodSeconds: 0
```
Comment (0xFelix) [inline tasks/create-vm-from-manifest/README.md:74]: Necessary for example?

File: tasks/create-vm-from-manifest/README.md [hunk 29]
```diff
@@ -17,7 +17,73 @@ This task creates a VirtualMachine from YAML manifest
...
+            volumes:
+            -   containerDisk:
+                    image: kubevirt/cirros-container-disk-demo
```
Comment (0xFelix) [inline tasks/create-vm-from-manifest/README.md:77]: Use one of the containerdisks provided by quay.io/containerdisks instead?

File: tasks/disk-virt-customize/README.md [hunk 30]
```diff
@@ -11,7 +11,36 @@ This task uses [virt-customize](https://libguestfs.org/virt-customize.1.html) to
...
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+    name: disk-virt-customize-taskrun
+spec:
+    params:
+    -   name: pvc
+        value: example-pvc
+    -   name: customizeCommands
+        value: 'update
```
Comment (0xFelix) [inline tasks/disk-virt-customize/README.md:25]: Use `|` multiline string instead?

File: tasks/disk-virt-sysprep/README.md [hunk 31]
```diff
@@ -9,10 +9,38 @@ This task uses [virt-sysprep](https://libguestfs.org/virt-sysprep.1.html) to run
...
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+    name: disk-virt-sysprep-taskrun
+spec:
+    params:
+    -   name: pvc
+        value: example-pvc
+    -   name: sysprepCommands
+        value: 'update
```
Comment (0xFelix) [inline tasks/disk-virt-sysprep/README.md:25]: Same

File: tasks/modify-data-object/README.md [hunk 32]
```diff
@@ -19,7 +19,35 @@ This task modifies a data object (DataVolumes or DataSources).
...
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+    name: modify-data-object-taskrun-resolver
+spec:
+    params:
+    -   name: waitForSuccess
+        value: 'true'
+    -   name: manifest
+        value: "apiVersion: cdi.kubevirt.io/v1beta1\nkind: DataVolume\nmetadata:\n
```
Comment (0xFelix) [inline tasks/modify-data-object/README.md:33]: Use multiline string?

File: templates/copy-template/readmes/README.md [hunk 33]
```diff
@@ -2,7 +2,7 @@
 
 #### Task is deprecated and will be removed in future versions.
 
-This task copies a template.
+This task copies an KubeVirt Virtual Machine template.
```
Comment (0xFelix) [inline templates/copy-template/readmes/README.md:5]:
```suggestion
This task copies a KubeVirt Virtual Machine template. 
```

File: tasks/modify-data-object/README.md [hunk 34]
```diff
@@ -19,7 +19,50 @@ This task modifies a data object (DataVolumes or DataSources).
...
+        -   name: kind
+            value: task
+        -   name: name
+            value: modify-data-object
+        -   name: version
+            value: v0.18.0
+        resolver: hub
+```
+
+In manifest parameter, you can use e.g. this manifest:
```
Comment (jcanocan) [inline tasks/modify-data-object/README.md:49]:
```suggestion
As an example for `manifest` parameter, you can use this DV definition:
```

File: templates/modify-data-object/readmes/README.md [hunk 35]
```diff
@@ -16,7 +16,13 @@ This task modifies a data object (DataVolumes or DataSources).
 
 ### Usage
 
-Please see [examples](examples) on how to modify data objects.
+Task run using resolver:
+```
+{{ task_run_resolver_yaml | to_nice_yaml }}```
+
+In manifest parameter, you can use e.g. this manifest:
```
Comment (jcanocan) [inline templates/modify-data-object/readmes/README.md:23]:
```suggestion
As an example for `manifest` parameter, you can use this DV definition:
```

**Comment by jcanocan** [comment]
Thanks! That's all on my side!
/lgtm

**Comment by 0xFelix** [comment]
Thanks!

/lgtm

**Pull request merged**